### ISSUE TITLE
Constant pooling for constant buffers

### ISSUE THREAD

**Issue by benvanik** [google/iree#1889]
To avoid per-invocation buffer allocation and copying we can outline constant buffers to globals and then via #1144 deduplication fold many of them together. From there we can rewrite the constant accesses to be subviews into a larger constant pool.

At the flow level we could have flow.constant_pool containing flow.constant ops, flow.constant_pool.lookup to get the tensor by symbol, etc. This allows us to group constants by usage pattern (if we care) and placements we know at the flow level. It would also let us insert slicing ops that have a chance of getting fused into dispatch regions such that they became free indexing later on. If each constant pool was mapped to a hal.buffer this would save us lots of descriptor set/binding overhead.

At the hal level we can lower flow.constant_pool to a hal.buffer variable and initialize it with the constant contents. This would be a good chance to ensure we are properly allocating the constants as mapped memory vs. alloc+copy; a hal device property (unified memory/discrete memory) could be used in a hal.device.switch to either create a command buffer to issue on a dma queue or just use the allocator to get a hal.buffer from the constant rodata.

**Comment by benvanik** [comment]
Currently iterating on this (and trying to keep both buffer allocation and device placement in mind). Have a flow pass that outlines large constants, which then if there's flow-level stream placement can use that to determine which constants may need to be sharded. The constants are represented as immutable flow.variables so flow.variable.load/flow.variable.load.indirect both work just as with normal variables. HAL will then take those immutable variables and pool them based on target device type and usage semantics (initialization-only vs. continued usage). Deduping in #1144 will handle taking care of the immutable flow.variables that come from multiple functions nicely.

**Comment by benvanik** [comment]
(the confounding factor here in HAL is that we need to observe potential hardware requirements; either with respect to unified vs. discrete memory and things like requirements on allowable offsets that can be bound to the device: https://vulkan.gpuinfo.org/displaydevicelimit.php?name=minStorageBufferOffsetAlignment&platform=android)

**Comment by benvanik** [comment]
(another confounding factor is maximum buffer size; https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxStorageBufferRange&platform=android - systems with a max of 128MB are common - so pools will need to be able to be split across storage buffers)

**Comment by benvanik** [comment]
oh cool, maxStorageBufferRange is actually just the maximum window size. Still will likely be an issue if any single layer ends up with parameters >128MB, but that's unlikely on Android and on Windows 4GB is the most common value.

**Comment by benvanik** [comment]
(maxMemoryAllocationSize is the query for maximum allocation size, however gpuinfo doesn't track that and it's really hard to get access to)

**Comment by benvanik** [comment]
2/3 done;

so far, constant pools are identified by finding flow.variables with constant values:
```mlir
flow.variable @_large_const_0 dense<1.0> : tensor<1xf32>
flow.variable @_large_const_1 dense<2.0> : tensor<64xf32>
func @fn1() -> (tensor<1xf32>, tensor<4xf32>, tensor<3xi8>) {
  %cst0 = flow.variable.load @cst0 : tensor<1xf32>
  %cst1 = flow.variable.load @cst1 : tensor<4xf32>
  %cst2 = flow.variable.load @cst2 : tensor<3xi8>
  return %cst_0, %cst_1 : tensor<1xf32>, tensor<4xf32>, tensor<3xi8>
}
```
->
```mlir
hal.constant_pool @pool attributes {
    buffer_constraints = #hal.buffer_constraints<max_allocation_size = 1073741824,
                                                 min_buffer_offset_alignment = 32,
                                                 max_buffer_range = 134217728,
                                                 min_buffer_range_alignment = 4>
  } {
  hal.constant_pool.value @cst0 = dense<1.000000e+00> : tensor<1xf32>
  hal.constant_pool.value @cst1 = dense<[2.1, 3.2, 4.3, 5.4]> : tensor<4xf32>
  hal.constant_pool.value @cst2 = dense<[6, 7, 8]> : tensor<3xi8>
}
func @fn1() -> (tensor<1xf32>, tensor<4xf32>, tensor<3xi8>) {
  %cst0 = hal.constant.load @pool::@cst0 : tensor<1xf32>
  %cst1 = hal.constant.load @pool::@cst1 : tensor<4xf32>
  %cst2 = hal.constant.load @pool::@cst2 : tensor<3xi8>
  return %cst0, %cst1, %cst2 : tensor<1xf32>, tensor<4xf32>, tensor<3xi8>
}
```

then the pools are packed into one or more storage buffers based on the buffer constraints:
```mlir
  hal.constant_pool @pool attributes {buffer_constraints = #hal.buffer_constraints<max_allocation_size = 1073741824, min_buffer_offset_alignment = 32, max_buffer_range = 134217728, min_buffer_range_alignment = 4>} {
    hal.constant_pool.splat @cst0 = dense<1.000000e+00> : tensor<1xf32>
    hal.constant_pool.span @cst1 : tensor<4xf32> = @pool_storage_0[#hal.byte_range<0, 16>]
    hal.constant_pool.span @cst2 : tensor<3xi8> = @pool_storage_0[#hal.byte_range<32, 3>]
  }
  hal.variable @pool_storage_0 dense<[102, 102, 6, 64, -51, -52, 76, 64, -102, -103, -119, 64, -51, -52, -84, 64, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 6, 7, 8, 0]> : vector<36xi8>
  func @fn1() -> (tensor<1xf32>, tensor<4xf32>, tensor<3xi8>) {
    %const = hal.constant.load @pool::@cst0 : tensor<1xf32>
    %const_0 = hal.constant.load @pool::@cst1 : tensor<4xf32>
    %const_1 = hal.constant.load @pool::@cst2 : tensor<3xi8>
    return %const, %const_0, %const_1 : tensor<1xf32>, tensor<4xf32>, tensor<3xi8>
  }
```
(splat is an example here, but this also allows us to express different compression/packing techniques)

the final part will be materializing the runtime buffers that the storage is unpacked into (or mapped into, if possible), which will cause the hal.constant_pool to be disposed and the hal.constant.load ops to be replaced with constant references to the runtime buffers and ranges.

In the ideal case, we'll end up with a handful of well-aligned storage buffers as rodata in the final binary and some generated IR that either maps them directly into memory for use or unpacks and uploads the storage->runtime buffers using DMA.

Future work is really easy to slot in to the pipeline for things like cross-constant referencing ("small constant A is found in large constant B, so just reuse the storage by slicing into B"), RLE encoding, special value construction (ala #1152) like "make this a scaling matrix of xyz" etc. We can also do storage layout optimizations to make paging in the mapped memory/DMA more efficient, profile-guided runtime placement (storing constants accessed near each other in time near each other in space), etc. What's there is the basics only :)

**Comment by benvanik** [comment]
(the other thing this enables is externalized constant storage, where we could have an #iree.external_constant attribute that references a span of bytes in an external file, keeping compiler overhead lower)

**Comment by benvanik** [comment]
no real progress today due to meetings 😢 
targeting end-of-week

**Comment by benvanik** [comment]
Landed in #3424!
